### ISSUE TITLE
deferred load breaks under unforked SDK + serialized compiliation

### ISSUE THREAD

**Issue by sigmundch** [dart-lang/sdk#41692]
Here is a repro:
**a.dart**
```dart
import 'b.dart' deferred as b;

main() => b.loadLibrary().then((_) => b.foo());
```

**b.dart**
```dart
class A<T> {
  final String key;
  const A(this.key);
}

const a1 = const A<String>('key1');
const a2 = const A<List<int>>('key2');

foo() {
  print(a1);
  print(a2);
  print(a1 == a2);
}
```

**single-step compile**:
```
dart2js --out=out.js --platform-binaries=$SDK/out/ReleaseX64NNBD/ a.dart
```
Will generate both the definition and references to C_A and C_A0 in out.js_1.part.js.

**serialized-step compile**:
```
dart2js --out=1.dill --platform-binaries=$SDK/out/ReleaseX64NNBD/ --cfe-only a.dart
dart2js --out=2.dill --write-data=2.data --platform-binaries=$SDK//out/ReleaseX64NNBD/ 1.dill
dart2js --out=out.js --read-data=2.data --platform-binaries=$SDK/out/ReleaseX64NNBD/ 2.dill
```

Will generate the definitions in out.js_1.part.js, but the references in out.js (main unit).

**Comment by sigmundch** [comment]
@fishythefish - I wont have much time to investigate today during the day - would you be able to look into it?

**Comment by fishythefish** [comment]
Sure, I'll give it a shot.

**Comment by rakudrama** [comment]
Adding `--enable-experiment=non-nullable` to the third command eliminates the difference.
I surmise that `2.dill` no longer feature-detects to the correct settings.
We could also store setting in the data.

**Comment by sigmundch** [comment]
:facepalm: - thanks @rakudrama, here is the fix https://dart-review.googlesource.com/c/sdk/+/147344

**Comment by sigmundch** [comment]
fixed in fb11d6399c7742333ab3ea9abe6d8ac88466dc4b
